Further improvements to synthtrace DSL and reintroducing workers to the CLI tool.

Mpdreamz opened the pull request:
## Refactor DSL
 
- Naming change from being span based to entity based. E.g `SpanIterable` => `EntityIterable`
- `concat` two streams together is now done through `merge` the resulting iterable round robins over all the streams it holds weighted by rate per minute (more on that later)
- `randomize()` on `Interval` allows you to quickly create intervals with variable yields.
- Included unit tests for the DSL behavior, ratePerMinute as well as weighted merging of entity generators.


## Rate per minute estimates.

Both `Interval` and `EntityIterable` implementations can now report an estimated event rate per minute. This is used in various places but most notably to estimate the range needed to satisfy `--maxDocs`. It's also used to automatically configure a proper segmentation of the range over a sufficient ammount of workers. 

```
$ node src/scripts/run src/scripts/scenarios/01_low_throughput.ts --local --clean --maxDocs 500000000
```

Will automatically infer a proper `from` and `to` range that will yield to number of documents we want from the scenario. It also automatically splits this range over N workers where the number of workers is calculated as well. It uses a `min(num_cores, 10)` as maximum. It will also ensure each worker does enough work so specifying `--maxDocs 1000` would never yield 10 workers as 1 would be sufficient.


## CLI improvements

- The return of the workers, this reintroduces workers to the CLI. Thanks to the `ratePerMinute()` estimates you can now target any scenario and the tool will automatically select an appropriate overal range, worker ranges + concurrency settings.
- Improves reporting e.g it now provides visual progress updates every `Math.min(2000000, maxDocs /20)` documents.
	- A summary for each worker is now printed at the end of the run as well.
- Introduces `--local` flag
	- For local development purposes this discovers both Elasticsearch and Kibana (with current basePath).
- An explicit `--kibana` flag also exists to accompay `--target`
- Will automatically install the appropriate APM integration package if either `--kibana` or `--local` are used.
	- When using `--cloudId` we still use our own migrate API to move over.
- Adds a `--gcpRepository` flag this is to aid with restoring data or snapshotting a large local dataset to share with others. See also https://github.com/elastic/kibana/pull/126938
- Added a `--maxDocConfidence` parameter, `ratePerMinute` works by peeking the first document of a stream. If your traces are more variable then that you can amplify/dampen the estimates using this parameter. 
- `--dryRun` will simply iterate over the `StreamProcessor`, useful to debug processor performance.
- The tool now assumes datastreams, you have to set `--forceLegacyIndices` to revert to old writeTargets behavior.
- The CLI now also discovers the running Elasticsearch version and makes sure the documents `StreamProcessor` yields have an equivalent `observer.version` so that APM's pipeline processors accept the source documents.
-

Mpdreamz committed:
Moves generator over to generators

* allow for lazy data generation
* feed lazy evaluation to a streamProcessor that does the aggregation.
* move entities over to use polymorphic this return types.
* introduces `SpanIterable` to aid with fluent construction

Mpdreamz committed:
Move to elasticsearch client bulk helper

and let it handle partitioning the data and concurrently index the data

Mpdreamz committed:
move new processors and hanlding into seperate files

Mpdreamz committed:
clean up StreamProcessor

Mpdreamz committed:
rename flatMap on SpanIterable to spans

Mpdreamz committed:
return datastreams always when calling getApmWriteTargets

Mpdreamz committed:
Add logger performance

Mpdreamz committed:
Update synthtrace ingestion to for 8.0.0

- Migrate cloud to managed out of the box
- Update to latest datastream mappings
- Move enrichment to StreamProcessor
- Instantiate services once, getCommonServices felt a bit service locator-ish

Mpdreamz committed:
Can now generate up to N spans

Mpdreamz committed:
Can now update the custom component templates

- For now we explicitly set number_of_shards to 1 (already the default)

Mpdreamz committed:
Remove check before swap, always swap

Mpdreamz committed:
update apm_api_integration usage of synthtrace lib

Mpdreamz committed:
fix compilation failures post-rebase

Mpdreamz committed:
make sure we generate transaction.root when calculating transaction metrics

Mpdreamz committed:
update timerange generation to generate bidirectionally properly again

Mpdreamz committed:
update EsLint errors

Commit:
commit using @elastic.co

Commit:
Address EsLint violations

Commit:
rename StackMonitoringFields back

Commit:
address TODO to move live indexation over to Syntthrace ES client

kibanamachine committed:
Merge branch 'main' into feature/synthtrace-stream-processing

Commit:
address post merge compilation error

Mpdreamz committed:
Introduce --dryRun and include order on SpanIterable.

This now allows `StreamProcessor` to take into account streams that are
generated in both ascending and descending order e.g when `--maxDocs` is
specified.

kibanamachine committed:
Merge branch 'main' into feature/synthtrace-stream-processing

Mpdreamz committed:
StreamProcessor improvements

- add logger so it can report to the standard out.
- If multiple streams are passed it now interleaves the streams into a single stream.
  This ensures no single stream has preference

Mpdreamz committed:
Fixes incorrect categorization of transaction metrics as applicaton metrics.

Mpdreamz committed:
staging various improvements

Mpdreamz committed:
Introduce rate_per_minute calculations on interval + streams

This allows merge() to proportionately yield from multiple streams and
run to better estimate ranges for min docs and parallelism

Mpdreamz committed:
Apply refactor to usages in tests and apm app

Mpdreamz committed:
Rename spans to generator

Since the DSL is not limited to generating spans

Mpdreamz committed:
Fix a bug in ratePerMinute calculation

Sub minute rates were normalized on the minute due to `Math.ceil` normalizing anything > 1 to 1.

Mpdreamz committed:
run improvements, introduces range estimations when using maxDocs

Mpdreamz committed:
add automatic workers to run

Mpdreamz committed:
make sure local development is also easy with synthtrace

Mpdreamz committed:
flip default to forceLegacyIndices instead of forcing datastreams

Mpdreamz committed:
update local_setup.md

Mpdreamz committed:
Cleanup

Mpdreamz committed:
Add option to register gcp repository, will go hand in hand with updates to `yarn es snapshot` for which a new PR will be created

Mpdreamz committed:
Make sure live data generation still works after all the refactoring.

Simplify support options for the CLI tool

Mpdreamz committed:
Merge branch 'main' into feature/synthtrace-stream-processing

Mpdreamz committed:
fix new usages of synthtrace DSL

Mpdreamz committed:
fix eslint error

Mpdreamz committed:
move latency api integration over to new DSL

Mpdreamz committed:
address linter issues

Mpdreamz committed:
Merge branch 'main' into feature/synthtrace-stream-processing

Mpdreamz committed:
post merge update of DSL

Mpdreamz committed:
Merge branch 'main' into feature/synthtrace-stream-processing

Mpdreamz committed:
fix linter errors

Mpdreamz committed:
Merge main into master

Mpdreamz committed:
Merge branch 'main' into feature/synthtrace-stream-processing-improvements-workers

elasticmachine commented:
Pinging @elastic/apm-ui (Team:apm)

kpatticha commented on packages/elastic-apm-synthtrace/src/lib/apm/client/apm_synthtrace_kibana_client.ts:
```diff
@@ -40,4 +40,36 @@ export class ApmSynthtraceKibanaClient {
+  }
+
+  async installApmPackage(kibanaUrl: string, version: string, username: string, password: string) {
+    const response = await fetch(kibanaUrl + '/api/fleet/epm/packages/apm/' + version, {
```
Does fleet _usually_ release the apm package  in correspondence to elasticsearch version? 
For example, `/api/fleet/epm/packages/apm/v8.2.0` doesn't exists instead is named `v8.2.0-dev3`

Mpdreamz replied:
It does but this needs to be HEAD aware where we are working ahead of the next release. 

Amended the PR with [`e436669` (#127257)](https://github.com/elastic/kibana/pull/127257/commits/e43666991bbf4ef5da7e8aa60e36c5104315bde2) to make the version selection of the package go through an actual resolving process.

kpatticha replied:
nice

Mpdreamz committed:
Merge branch 'main' into feature/synthtrace-stream-processing-improvements-workers

Mpdreamz committed:
post merge update of DSL

Mpdreamz committed:
post merge linter updates

Mpdreamz committed:
add discovery of latest valid package version for the currently running kibana version

Mpdreamz committed:
declare semver in bazel build

Mpdreamz committed:
fix linter errors

Mpdreamz committed:
update toDocument unit tests

Mpdreamz committed:
update static invocation of toDocument

Mpdreamz committed:
Introduce flag to refresh after index on ApmSynthtrace client

Both API and Cyprus tests rely on this behavior.

Added ApmSynthtraceEsClientOptions to reflect the new defaulting to datastreams more clearly.

You have to opt-in to legacy indices now.

Mpdreamz committed:
precalculate intervals also when --maxDocs is not specified.

Mpdreamz committed:
remove clientWorkers argument

Mpdreamz committed:
update README with new options

Mpdreamz committed:
erroneous check for maxDocs when processing worker messages

Mpdreamz committed:
add naive --target inference if only --kibana to a cloud url is specified

Mpdreamz committed:
Merge branch 'main' into feature/synthtrace-stream-processing-improvements-workers

Mpdreamz committed:
Merge branch 'main' into feature/synthtrace-stream-processing-improvements-workers

# Conflicts:
#	packages/elastic-apm-synthtrace/src/index.ts
#	x-pack/plugins/apm/dev_docs/local_setup.md
#	x-pack/plugins/apm/ftr_e2e/cypress/plugins/index.ts

Mpdreamz committed:
update API test to new DSL post merge

Mpdreamz committed:
update API test to new DSL post merge

Mpdreamz committed:
export EntityArrayIterable directly

dgieselaar reviewed (approved):
Approving to get this in, let's iterate on it once the other PR is in as well

kibanamachine commented:
Friendly reminder: Looks like this PR hasn’t been backported yet.
To create backports run `node scripts/backport --pr 127257` or prevent reminders by adding the `backport:skip` label.

kibanamachine commented:
Friendly reminder: Looks like this PR hasn’t been backported yet.
To create backports run `node scripts/backport --pr 127257` or prevent reminders by adding the `backport:skip` label.

kibanamachine commented:
Friendly reminder: Looks like this PR hasn’t been backported yet.
To create backports run `node scripts/backport --pr 127257` or prevent reminders by adding the `backport:skip` label.

kibanamachine commented:
Friendly reminder: Looks like this PR hasn’t been backported yet.
To create backports run `node scripts/backport --pr 127257` or prevent reminders by adding the `backport:skip` label.

kibanamachine commented:
Friendly reminder: Looks like this PR hasn’t been backported yet.
To create backports run `node scripts/backport --pr 127257` or prevent reminders by adding the `backport:skip` label.

kibanamachine commented:
Friendly reminder: Looks like this PR hasn’t been backported yet.
To create backports run `node scripts/backport --pr 127257` or prevent reminders by adding the `backport:skip` label.

kibanamachine commented:
Friendly reminder: Looks like this PR hasn’t been backported yet.
To create backports run `node scripts/backport --pr 127257` or prevent reminders by adding the `backport:skip` label.

kibanamachine commented:
Friendly reminder: Looks like this PR hasn’t been backported yet.
To create backports run `node scripts/backport --pr 127257` or prevent reminders by adding the `backport:skip` label.
